kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/RecordConverter.java

```diff
@@ -0,0 +1,508 @@
+    }
+    throw new RuntimeException("Cannot convert time: " + value);
+  }
+
+  protected Temporal convertTimestampValue(Object value, TimestampType type) {
+    if (type.shouldAdjustToUTC()) {
+      return convertOffsetDateTime(value);
+    }
+    return convertLocalDateTime(value);
+  }
+
+  @SuppressWarnings("JavaUtilDate")
+  private OffsetDateTime convertOffsetDateTime(Object value) {
+    if (value instanceof Number) {
+      long millis = ((Number) value).longValue();
+      return DateTimeUtil.timestamptzFromMicros(millis * 1000);
+    } else if (value instanceof String) {
+      return parseOffsetDateTime((String) value);
+    } else if (value instanceof OffsetDateTime) {
+      return (OffsetDateTime) value;
+    } else if (value instanceof LocalDateTime) {
+      return ((LocalDateTime) value).atOffset(ZoneOffset.UTC);
+    } else if (value instanceof Date) {
+      return DateTimeUtil.timestamptzFromMicros(((Date) value).getTime() * 1000);
+    }
+    throw new RuntimeException(
+        "Cannot convert timestamptz: " + value + ", type: " + value.getClass());
+  }
+
+  private OffsetDateTime parseOffsetDateTime(String str) {
+    String tsStr = ensureTimestampFormat(str);
+    try {
+      return OFFSET_TS_FMT.parse(tsStr, OffsetDateTime::from);
+    } catch (DateTimeParseException e) {
+      return LocalDateTime.parse(tsStr, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
+          .atOffset(ZoneOffset.UTC);
+    }
+  }
+
+  @SuppressWarnings("JavaUtilDate")
+  private LocalDateTime convertLocalDateTime(Object value) {
+    if (value instanceof Number) {
+      long millis = ((Number) value).longValue();
+      return DateTimeUtil.timestampFromMicros(millis * 1000);
+    } else if (value instanceof String) {
+      return parseLocalDateTime((String) value);
+    } else if (value instanceof LocalDateTime) {
+      return (LocalDateTime) value;
+    } else if (value instanceof OffsetDateTime) {
+      return ((OffsetDateTime) value).toLocalDateTime();
+    } else if (value instanceof Date) {
+      return DateTimeUtil.timestampFromMicros(((Date) value).getTime() * 1000);
+    }
+    throw new RuntimeException(
+        "Cannot convert timestamp: " + value + ", type: " + value.getClass());
+  }
+
+  private LocalDateTime parseLocalDateTime(String str) {
+    String tsStr = ensureTimestampFormat(str);
+    try {
+      return LocalDateTime.parse(tsStr, DateTimeFormatter.ISO_LOCAL_DATE_TIME);
+    } catch (DateTimeParseException e) {
+      return OFFSET_TS_FMT.parse(tsStr, OffsetDateTime::from).toLocalDateTime();
+    }
+  }
+
+  private String ensureTimestampFormat(String str) {
+    String result = str;
+    if (result.charAt(10) == ' ') {
+      result = result.substring(0, 10) + 'T' + result.substring(11);
+    }
+    if (result.length() > 22 && result.charAt(19) == '+' && result.charAt(22) == ':') {
+      result = result.substring(0, 19) + result.substring(19).replace(":", "");
+    }
+    return result;
+  }
```

Comment:
This was requested by users and it seemed reasonable to permit some well known variations of the ISO formats (e.g. space instead of `T` or offsets with a `:`).